### DISCUSSION TITLE
Can I make a callback (e.g. on_click) execute a sequence of functions?

### DISCUSSION THREAD

**Discussion by duane-space** [zauberzeug/nicegui#1220] [Q&A]
### Question

Right now there are a lot of places I do something like:

```py
button.on('click', functools.partial(function, *arguments))
```

I find often a lot of the arguments end up being elements (to do things such as disable buttons while the function runs, or add a loading status to a table, etc). I've been trying to clean up my code and one objective has been to reduce the amount of passing of elements I do. One way I have found success was using refreshable, which has been great.

But there are other places this didn't help me get away from it, and I think one way to help would be being able to call a series of functions in a row on an event. Something like:

```py
button.on('click',[button.disable(), functools.partial(function, *arguments), button.enable()])
```

Is there already a good way to do this I don't know about?

_Side note: I actually don't remember exactly why I use functools so much... if I recall it had something to do with converting lambda functions to work asynchronously?_

**Comment by falkoschindler** [comment]
Hi @duane-impulse!

The straight forward approach would be to define a separate event handler:

```py
async def handle_click(arguments):
    button.disable()
    await some_function(arguments)
    button.enable()

button.on('click', lambda: handle_click(arguments))
```

Note that the lambda function returns an Awaitable, which is automatically awaited by NiceGUI's event mechanism.

For non-async functions you can also use this trick to wrap multiple function calls in one lambda expression:
```py
button.on('click', lambda: (
    some_function(),
    some_other_function(),
))
```

The lambda expression will evaluate the tuple, and therefore it evaluates every function call it contains.

**Reply by duane-space** [reply]
Thanks! I have all my business logic as imports and got stuck thinking in that paradigm, vs defining handlers in scope so they can access the elements directly

**Reply by duane-space** [reply]
Actually - follow up question!

In the first example, I would need to specifically make a handler for every button I want to exhibit this behavior. Is it possible to do this once generically and use for multiple buttons? The first thing that comes to my mind would be as a decorator for the functions where I do other business logic. I think this would work, but the way I'm thinking about it would head me back towards passing UI elements around... although maybe there is a better way. 

I guess another idea could be a custom element that inherits from button and overrides the click event, but it does seem to me like this simple case (disabling a button until the on click callback completes) would be a common scenario.

I do have other (more complex, less common) use cases where a specific in-scope handler is likely the best solution.

**Reply by duane-space** [reply]
Just noticed this is related to some discussion in this pr...

https://github.com/zauberzeug/nicegui/pull/1238

I think the auto_disable is a good idea. I can probably see arguments for defauling to either True or False.
